### PR TITLE
feat: allow php 8

### PR DESCRIPTION
**Pull request by marko-ilic**

### CODE REVIEW

File: src/AbstractSnapshot.php [hunk 1]
```diff
@@ -109,7 +109,7 @@ final public function getDataKeys()
      *
      * @codeCoverageIgnore
      */
-    final private function __clone()
+    private function __clone()
```
Comment (Taluu) [inline src/AbstractSnapshot.php]: why ?

File: .gitignore [hunk 2]
```diff
@@ -10,4 +10,6 @@ coveralls-upload.json
 vendor
 composer.lock
 bin
+.phpunit.result.cache
+.idea
```
Comment (Taluu) [inline .gitignore]: don't, this should be in *your* global .gitignore

File: composer.json [hunk 3]
```diff
@@ -32,7 +32,7 @@
     },
 
     "require": {
-        "php": "^5.4 || ^7.0",
+        "php": "^7.3 || ^7.4 || ^8.0",
```
Comment (Taluu) [inline composer.json]:
```suggestion
        "php": "^7.3 || ^8.0",
```

Is enough

File: README.md [hunk 4]
```diff
@@ -12,7 +12,7 @@ Totem
        \\\\////
         |.)(.|
         | || |   Changeset calculator between two state of a data
-        \(__)/   Requires PHP 5.4 ; Compatible PHP 5.5, PHP 5.6, PHP 7 and HHVM
+        \(__)/   Compatibile with PHP 7.2, 7.3, 7.4 or 8
```
Comment (Taluu) [inline README.md]: according to the changes in composer, 7.2 is not supported anymore

File: test/AbstractSnapshotTest.php [hunk 5]
```diff
@@ -53,32 +51,29 @@ public function existsProvider()
                 ['bar', false]];
     }
 
-    /**
-     * @expectedException        BadMethodCallException
-     * @expectedExceptionMessage A snapshot is frozen by nature
-     */
     public function testOffsetUnset()
     {
+        $this->expectException(\BadMethodCallException::class);
```
Comment (Taluu) [inline test/AbstractSnapshotTest.php]: add uses in top please

File: test/Snapshot/ArraySnapshotTest.php [hunk 6]
```diff
@@ -42,7 +41,9 @@ public function providerCompare()
      */
     public function testDeepConstructor($value)
     {
-        new ArraySnapshot(['foo' => $value]);
+        $arraySnapshot = new ArraySnapshot(['foo' => $value]);
+
+        self::assertSame(['foo' => $value], $arraySnapshot->getRawData());
```
Comment (Taluu) [inline test/Snapshot/ArraySnapshotTest.php]: IIRC, this test was just to ensure the constructor didn't throw anything, and it looks useless ot test then the raw data.

But I guess PHPUnit was reporting this test as incomplete or something ? Wasn't there a method or a call that tells phpunit to ignore such errors ?

**Review by Taluu — CHANGES_REQUESTED** [review]
A few things to workout, especially composer stuff.

The other looks fine

File: src/AbstractSnapshot.php (hunk 1 above)
Comment (marko-ilic) [reply src/AbstractSnapshot.php]: Private method can not be final

Comment (Taluu) [reply src/AbstractSnapshot.php]: Forgot about that. Carry on. :D

Comment (Taluu) [reply src/AbstractSnapshot.php]: (You can remove this method then)

File: test/Snapshot/ArraySnapshotTest.php (hunk 6 above)
Comment (marko-ilic) [reply test/Snapshot/ArraySnapshotTest.php]: IMO test should perform assertion(s) (or expect exception(s)). Otherwise they are pointless, right?

Comment (marko-ilic) [reply test/Snapshot/ArraySnapshotTest.php]: And yes, you're right, tests without assertions (or expected exceptions) are marked as risky, that's why I modified this one.

**Comment by healerz** [comment]
Hi @Taluu 
Any chance to merge and release this PR soon? 🙂

**Review by Taluu — APPROVED** [review]
Looks good to me ! :}

**Comment by Taluu** [comment]
Done : https://github.com/Wisembly/Totem/releases/tag/v1.5.0 :tada:

**Comment by healerz** [comment]
@Taluu Thank you 👍

**Pull request merged**